pkg/controller/certificates/util.go

```diff
@@ -83,25 +83,34 @@ func certificateMatchesSpec(crt *v1alpha2.Certificate, key crypto.Signer, cert *
 	// TODO: add checks for IsCA field
 
 	// check if the private key is the corresponding pair to the certificate
+
 	matches, err := pki.PublicKeyMatchesCertificate(key.Public(), cert)
 	if err != nil {
 		errs = append(errs, err.Error())
 	} else if !matches {
 		errs = append(errs, fmt.Sprintf("Certificate private key does not match certificate"))
 	}
 
-	// validate the common name is correct
-	expectedCN := pki.CommonNameForCertificate(crt)
-	if expectedCN != cert.Subject.CommonName {
-		errs = append(errs, fmt.Sprintf("Common name on TLS certificate not up to date: %q", cert.Subject.CommonName))
+	// If CN is set on the resource then it should exist on the certificate as
+	// the Common Name or a DNS Name
+	expectedCN := crt.Spec.CommonName
+	gotCN := append(cert.DNSNames, cert.Subject.CommonName)
+	if len(expectedCN) > 0 && !util.Contains(gotCN, expectedCN) {
+		errs = append(errs, fmt.Sprintf("Common Name on TLS certificate not up to date (%q): %s",
+			expectedCN, gotCN))
```

Comment:
Would `k8s.io/apimachinery/pkg/util/sets.String` be easier to work with here...? It has `Insert`, `Delete`, `Has`, `Intersection`, `IsSuperset` methods etc